### ISSUE TITLE
Final v0.9.0 Tasks (and Namespacing)

### ISSUE THREAD

**Issue by jdantonio** [ruby-concurrency/concurrent-ruby#324]
**TL;DR: I believe we are ready to lock in the public API, release v0.9.0, and begin work on v1.0.**

With PRs #322 and #323 (and the BiasedConditionVariable [requested by the Rails team](https://github.com/rails/rails/pull/14938/files#r30142179)) I believe we are ready to release v0.9.0. This issue is to discuss final thoughts before that release.

When we first began discussing v0.9.0, in Issue #257, our goal was to stabilize the API so that we could focus on a stable, reliable, performant 1.0 release. Although *many* of the changes were of huge value to our users (creating an Edge gem, the common synchronization layer, bug fixes, etc.) the scope of the release began to grow. Many of the changes we undertook were internal changes only that had no functional benefit (newer pattern for runtime-specific subclasses, submodules for private abstractions). As a result we pushed the release date numerous times. I believe this was a mistake. We lost sight of the fact that "shipped" software is *always* better than "perfect" software.

One of the major discussions concerned the "namespace" modules for internal classes (see #261). At issue was whether or not we should make internal changes for greater consistency and to reflect common Ruby conventions. Although these suggestions had merit, they also represented a significant effort and a departure from what was already working an in production for many if our users. I began to reorganize *some* of what was discussed in PR #304 but had to abandon that effort due to merge conflicts. Some of that work was revisited in PRs #316, #317, #319, #320, and #321.

At this time I believe very strongly that we should release v0.9.0, move forward with the original plan for v1.0, and revisit namespace reorganization after v1.0.

According to the rules of Semantic Versioning, only changes to the *public* API must result in changes to the major and minor version numbers. Changes that are *private* are exempt. The namespace changes we discussed in #261 are *private* changes. We can make those changes, slowly and deliberately, once v1.0 is released an not violate the rules of Semantic Versioning. To facilitate this I have tagged all internal, private abstractions with the appropriate Yardoc tag (see PR #322). This PR (and any subsequent updates) should satisfy the requirements of Semantic Versioning.

Additional thoughts? Agree/disagree? @pitr-ch @mighe @chrisseaton @rkday @obrok @lucasallan @headius

**Comment by pitr-ch** [comment]
I think we should have another pre-release with the complete change-log (so users like @SebastianEdwards have time to try it before we release). Since we are releasing edge for the first time I would also like #314 to be merged (still few thinks to do there) before pre2.

We should also talk about how do we edge releases in future. Since it shares README file and some other pieces I think we should always bump patch (or minor?) version of core and do appropriate semantic bump on edge.

When would you like to add the BiasedCondition 0.9 or 1.0? It will require more time to understand it and to decide if and how do we add it to `concurrent-ruby`.

Nice work on the #316, #317, #319, #320, and #321! Before we release I would like to do a pass on whole public API if there is something we should consider changing, I'll comment here again after it's done. I am fine with doing rest of the changes on private parts later (preferably between 0.9-1.0 though).

My todos:
- [ ] Changelog updates `0.9.pre2`
- [ ] Finish and merge #314 `0.9.pre2`
- [ ] go through whole public API `0.9.pre2`
- [ ] resolve BiasedCondition `??`

**Comment by jdantonio** [comment]
In no particular order:

* I'm comfortable doing another pre-release of 0.9.
  - We should discuss #326 and discuss a final solution.
  - We can certainly merge #314, too, but I'd like to get that done as soon as reasonably possible. (Any idea how long that may take?)
* We can wait until 1.0 or later for BiasedCondition. I didn't look at the feature request too closely, I just assumed (perhaps incorrectly) that it wouldn't be difficult. We should definitely wait if necessary.
* I'd like to freeze the current feature set, unless there is something absolutely critical still outstanding. (I've probably been the worst about adding new features into this release.)
* I like the idea of releasing Edge in conjunction with Core, the same way we plan to release Ext and Core simultaneously. We can document something in the README later.
* We definitely need more people to review the private/public API. As we discussed in #322, I want to revisit the Yardoc tags with an eye to generating the best documentation.
* I'd like some help updating the CHANGELOG in PR #323.

**Comment by sschepens** [comment]
Are you planning on pushing pre-releases to rubygems? I wish to test them

**Comment by jdantonio** [comment]
@sschepens Yes, I will be publishing `v0.9.0.pre2` this weekend. Are you OK downloading the gem files from GitHub, or do you need them pushed to Rubygems?

**Comment by sschepens** [comment]
@jdantonio it would be great if you could push them to Rubygems, i'm developing a gem that would be using concurrent-ruby as a dependecy so i need it to be in Rubygems :)

**Comment by jdantonio** [comment]
@sschepens :cool: I'll push to Rubygems and make a note on this thread when I build the next pre-release. Do you mind if I ask what you are building? I'm very curious to see how others use this gem.

**Comment by sschepens** [comment]
I'm building an Amazon SQS message processor, it's based on [shoryuken](https://github.com/phstc/shoryuken) which uses Celluloid. I forked it because i was having memory leaks and wanted to change some things, I tried to use Celluloid but was having leak issues too and some other problems.
So I decided to give concurrent-ruby a try!

**Comment by jdantonio** [comment]
Awesome! Let us know how it goes and if we can do anything to help.

**Comment by pitr-ch** [comment]
@sschepens: exactly what @jdantonio said. Any feedback is very valuable and appreciated.

**Comment by sschepens** [comment]
@jdantonio @pitr-ch thanks!
I know Actors are and Edge feature but their documentation is really lacking, as well as Behaviors.
Also i'm having an 'issue' when an exception is thrown in an Actor pool.
A simple failing actor just gets reset on failure:
```ruby
class FailingActor < Concurrent::Actor::RestartingContext
  def on_message(msg)
    fail
  end
  def on_event(event)
    puts "event: #{event}"
  end
end

fa = FailingActor.spawn name: :test, supervise: true, args: []
fa << :test
#event:
#event: resetting
#event: reset
```

But the same actor in a a Pool gets into a reset loop:
```ruby
class FailingWorker < Concurrent::Actor::Utils::AbstractWorker
  def work(msg)
    fail
  end

  def on_event(event)
    puts "event: #{event}"
  end
end

pool = Concurrent::Actor::Utils::Pool.spawn! :pool, 5 do |balancer, index|
  FailingWorker.spawn name: "test_#{index}".to_sym, supervise: true, args: [balancer]
end

pool << :test

#event: 
#event: resetting
#event: reset
#event: 
#event: resetting
#event: reset
#event: 
#event: resetting
#event: reset
#event: 
#event: resetting
#event: reset
#event: 
#event: resetting
#event: reset
#event: 
#event: resetting
#event: reset
#...
``` 
I don't know if this is an expected behavior.

**Comment by jdantonio** [comment]
@pitr-ch I would like to release v0.9.0 this weekend. It seems like all the known bug have been addressed. Thoughts?

**Comment by pitr-ch** [comment]
@jdantonio I would be more comfortable with one more pre release to verify that all the fixes are working well for our users, and if they are, release next week. There is also https://github.com/ruby-concurrency/concurrent-ruby/issues/354 which I would like to fix before 0.9 even though it's kind a minor problem.

**Comment by jdantonio** [comment]
[Done](https://github.com/ruby-concurrency/concurrent-ruby/releases/tag/v0.9.0.pre3)

**Comment by jdantonio** [comment]
@pitr-ch @iNecas How does it look? Are we ready for 0.9.0 to be released this weekend?

**Comment by iNecas** [comment]
:+1: I hit no issues besides the minor one with the `require 'monitor'` this week. From API point of view there is all we need, and makes sense to give this to even more folks and address potential issues as they appear.

**Comment by pitr-ch** [comment]
i'll push #354 tomorrow and we'll be ready.

**Comment by iNecas** [comment]
Travis is also happy about the state of art https://travis-ci.org/Dynflow/dynflow/builds/68624765

**Comment by pitr-ch** [comment]
@iNecas nice. Could you also start running test on JRuby? That would be a big help!

**Comment by pitr-ch** [comment]
Unfortunately there are some remaining issues to be solved in #357 related #354.

**Comment by iNecas** [comment]
@pitr-ch planning to do that  (ETA in two weeks), I tried some time ago (and already filled issues accordingly https://github.com/ruby-concurrency/concurrent-ruby/issues/248). Now that it's fixed, I can give it another try.

Btw. what do you thing about an idea of including the Dynflow test suite in the concurrent ruby PR process (as 'allowed-fail' at least in the beginning). As we build the layer on top of concurrent ruby,
we test its features quite extensively, so it could be used as some kind of "system test", and also could help with identifying backward compatibility issues etc.

One good thing we do (IMO) is making sure, that every future in our tests gets resolved at some point: in concurrent tests this proved to be very helpful indicator of healthy code: often the tests are green, but due to some issue something gets oncompleted, which in production can lead to deadlocks, memory leaks etc.

Just an idea

**Comment by pitr-ch** [comment]
@iNecas Great thanks.

Definitely :+1: to all you've said, I have it on my todolist but I didn't had a time to try add dynflow test suite. If I may be so bold, would you have time to look into that? 

The future completion check could be done as utility helpers so any user of `concurrent-ruby` can use it in tests or even in production to hunt down bugs. That would be a great addition to `concurrent-ruby`, I look into porting what we had in Dynflow to `concurrent-ruby`.

**Comment by jdantonio** [comment]
How long does the Dynflow test suite take to run? I like the idea of a sanity check, but our test suite is painfully slow already. I'd hate to make PRs even slower. If the Dynflow test suite is slow, too, perhaps we can limit those runs to when we merge to master.

**Comment by iNecas** [comment]
It's 1:20 in average. Also, not all the tests use the concurrency futures
(such as persistence tests), in some cases, we test different variants of
connector (so basically running the same set of tests twice using different
drivers). Therefore, we could find a set of tests that use the
concurrent-ruby the most and make that as separate rake task. I think that
could take us under one minute, maybe even more.

What times would be acceptable?

-- Ivan

**Comment by jdantonio** [comment]
@ianks I don't think that another minute or so for the additional tests should be a problem so long as we aren't running them against too many runtimes. Perhaps we could run the Dynflow tests against the latest MRI when PRs are created then run against all supported platforms when merging to master?

**Comment by pitr-ch** [comment]
For me the test run-time on Travis is not concern until it takes hours. So I am for adding as many tests as it makes sense. At least CRuby 2.x and JRuby though.

**Comment by pitr-ch** [comment]
I would rather run all tests on PRs than having additional tests on master to avoid broken master.

**Comment by jdantonio** [comment]
Hours? You are *far* more patient than I am! :-)

**Comment by jdantonio** [comment]
If Dynflow test only add additional coverage of a small portion of concurrent-ruby then the added value of adding those tests to this CI build is proportionally small. Especially considering the generally high level of code coverage we already have. It's a nice bonus but not critical. And I wouldn't consider a problem uncovered by Dynflow to indicate a "broken" master, just an uncovered bug. A bug which should immediately result in the creation of an Issue which gets fixed with tests.

Running the Dynflow tests during our CI build doesn't alleviate anyone of the responsibility for creating thorough and robust tests in this project. So running the Dynflow tests on every CI build should never be necessary or critical.

**Comment by pitr-ch** [comment]
That's true it covers only actors, new futures, and pools (@iNecas please correct me if I am wrong). We could contact other users of concurrent-ruby to find more OSS projects we could run. You are right it does not indicate broken, I would have liked to have it simple green/red but there may be other issues like dynflow's (or other OSS project's) tests temporarily broken which would block us. So having it as allowed failure on PRs seems as a best option to me. Do we get an email about failure in allowed-failure-test?

**Comment by jdantonio** [comment]
There are no more open PRs and the new build task for building the Windows binaries works great. I'm ready to release core and ext v0.9.0, edge 0.1.0. I should have free time this evening. If anyone has any concerns about this release, please let me know ASAP.

I also plan to release [ref v2.0](https://github.com/ruby-concurrency/ref) tonight or this weekend, unless anyone has concerns.

Thank you to everyone, especially @pitr-ch, for all the hard work on this release! I'm *very* excited about the direction we are heading. This release--and the upcoming 1.0--are going to be HUGE!

Big thanks also to @lucasallan who did all the work on Ref 2.0.

Thank you!

**Comment by pitr-ch** [comment]
@jdantonio And big thanks to you too! Please go ahead and release. No concerns on mi side.

**Comment by jdantonio** [comment]
Well, crap. It may take me longer to release than anticipated. I've discovered a circular reference that doesn't cause a problem when running tests, but raises an exception when running the build tests. `concurrent/utility/et_exit` requires `concurrent/concern/logging` which requires `concurrent/configuration` which requires `concurrent/utility/at_exit`. I'll release once I figure it out.

**Comment by jdantonio** [comment]
Done!
